finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+    }
+
+    fn handle_broadcast(
+        &mut self,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> (Vec<Multiaddr>, Option<(DiscoveryMessage, DataCommand)>) {
+        let addresses = self.handle_authentication(authentication.clone(), handler);
+        if addresses.is_empty() {
+            return (addresses, None);
+        }
+        let node_id = authentication.0.creator();
+        let rebroadcast = match self.last_broadcast.get(&node_id) {
+            Some(instant) => time::Instant::now() > *instant + self.cooldown,
+            None => true,
+        };
+        let message = match rebroadcast {
+            true => {
+                self.last_broadcast.insert(node_id, time::Instant::now());
+                Some(authentication_broadcast(authentication))
+            }
+            false => None,
+        };
+        (addresses, message)
+    }
+
+    fn create_response(
+        &mut self,
+        requester_id: NodeIndex,
+        node_ids: Vec<NodeIndex>,
+        handler: &mut SessionHandler,
+    ) -> Option<(DiscoveryMessage, DataCommand)> {
+        let requested_authentications = self
+            .requested_authentications
+            .entry(requester_id)
+            .or_default();
+        requested_authentications.extend(node_ids);
+        if let Some(instant) = self.last_response.get(&requester_id) {
+            if time::Instant::now() < *instant + self.cooldown {
+                return None;
+            }
+        }
+        let peer_id = match handler.peer_id(&requester_id) {
+            Some(peer_id) => peer_id,
+            None => return None,
+        };
+        let authentications: Vec<_> = requested_authentications
+            .iter()
+            .filter_map(|id| handler.authentication_for(id))
+            .collect();
+        if authentications.is_empty() {
+            None
+        } else {
+            self.last_response
+                .insert(requester_id, time::Instant::now());
+            self.requested_authentications.remove(&requester_id);
+            Some(response(authentications, peer_id))
+        }
+    }
+
+    fn handle_request(
+        &mut self,
+        node_ids: Vec<NodeIndex>,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> (Vec<Multiaddr>, Option<(DiscoveryMessage, DataCommand)>) {
+        let node_id = authentication.0.creator();
+        let addresses = self.handle_authentication(authentication, handler);
+        if addresses.is_empty() {
+            return (addresses, None);
+        }
+        (addresses, self.create_response(node_id, node_ids, handler))
+    }
+
+    fn handle_response(
+        &mut self,
+        authentications: Vec<Authentication>,
+        handler: &mut SessionHandler,
+    ) -> Vec<Multiaddr> {
+        authentications
+            .into_iter()
+            .flat_map(|authentication| self.handle_authentication(authentication, handler))
```

Comment:
should we remove the duplicates?

Reply:
Shouldn't really matter, we might return the same address many times for other reasons, so I'm counting on the idempotence of `add_reserved` anyway.